service_extensions/callouts/add_header/server.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2023 Google LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+# Example external processing server
+----
+This server does two things:
+ When it gets request_headers, it replaces the Host header with service-extensions.com
+  and resets the path to /.
+ When it gets response_headers, it adds a "hello: service-extensions" response header.
+"""
+from concurrent import futures
+from http.server import BaseHTTPRequestHandler, HTTPServer
+import logging
+
+from typing import Iterator, List
+
+import grpc
+
+from grpc import ServicerContext
+
+import _credentials
+import service_pb2
+import service_pb2_grpc
+
+logger = logging.getLogger()
+logger.setLevel("INFO")
+
+# Backend services on GCE VMs, GKE and hybrid use this port.
+EXT_PROC_SECURE_PORT = 443
+# Backend services on Cloud Run use this port.
+EXT_PROC_INSECURE_PORT = 8080
+# Cloud health checks use this port.
+HEALTH_CHECK_PORT = 80
+
+
+# Returns an ext_proc HeadersResponse for adding a list of headers.
+# clear_route_cache should be set to influence service selection for route
+# extensions.
+def add_headers_mutation(
+    headers: List[tuple[str, str]], clear_route_cache: bool = False
+) -> service_pb2.HeadersResponse:
+    response_header_mutation = service_pb2.HeadersResponse()
+    response_header_mutation.response.header_mutation.set_headers.extend(
+        [
+            service_pb2.HeaderValueOption(
+                header=service_pb2.HeaderValue(key=k, raw_value=bytes(v, "utf-8"))
+            )
+            for k, v in headers
+        ]
+    )
+    if clear_route_cache:
+        response_header_mutation.response.clear_route_cache = True
+    return response_header_mutation
+
+
+class CalloutProcessor(service_pb2_grpc.ExternalProcessorServicer):
+    def Process(
+        self,
+        request_iterator: Iterator[service_pb2.ProcessingRequest],
+        context: ServicerContext,
+    ) -> Iterator[service_pb2.ProcessingResponse]:
+        for request in request_iterator:
+            print(request)
```

Comment:
question: is this for debugging? If so, we should probably remove before publishing.

Reply:
Yes. We can clear it.